File: tests/client_test.py
import json
import threading
import unittest
from http.server import BaseHTTPRequestHandler, ThreadingHTTPServer
from urllib.error import HTTPError
from expectbox_agents import ExpectboxProject, ExpectboxAgent

KEY = 'exp_' + 'p' * 43
INBOX = '12345678-1234-1234-1234-123456789abc'

class ClientTest(unittest.TestCase):
    def test_contract(self):
        seen = []
        class Handler(BaseHTTPRequestHandler):
            def log_message(self, *args): pass
            def do_POST(self):
                seen.append((self.path, self.headers, self.rfile.read(int(self.headers['Content-Length']))))
                self.send_response(200); self.end_headers()
                self.wfile.write(json.dumps({'id': INBOX}).encode())
            def do_GET(self):
                seen.append((self.path, self.headers, None))
                self.send_response(307); self.send_header('Location', '/no'); self.end_headers()
        server = ThreadingHTTPServer(('127.0.0.1', 0), Handler)
        thread = threading.Thread(target=server.serve_forever, daemon=True); thread.start()
        try:
            client = ExpectboxProject(KEY, f'http://127.0.0.1:{server.server_port}')
            for _ in range(2): client.create_inbox({'name': 'Test', 'username': 'test'}, 'saved-key')
            self.assertEqual(seen[0][0], '/api/agents/v1/project/inboxes')
            self.assertEqual(seen[0][1]['Idempotency-Key'], 'saved-key')
            self.assertEqual(seen[0][2], seen[1][2])
            client.create_inbox_key(INBOX, 'Runtime', ['messages:read'])
            with self.assertRaises(HTTPError): client.project()
            self.assertEqual(len(seen), 4)
            with self.assertRaises(ValueError): client.create_inbox({}, '')
            with self.assertRaises(ValueError): client.create_inbox_key('../x', 'Runtime', ['messages:read'])
            with self.assertRaises(ValueError): ExpectboxAgent(KEY)
        finally:
            server.shutdown(); server.server_close(); thread.join()

if __name__ == '__main__': unittest.main()

File: tests/client.test.mjs
import { test } from "node:test";
import assert from "node:assert/strict";
import { createServer } from "node:http";
import { ExpectboxProject, ExpectboxAgent } from "expectbox-agents";
const id = "12345678-1234-1234-1234-123456789abc";
test("installed SDK separates credentials and preserves provisioning identity", async (t) => {
  const received = [];
  const server = createServer(async (req, res) => {
    let body = "";
    for await (const chunk of req) body += chunk;
    received.push({
      url: req.url,
      headers: req.headers,
      body,
      method: req.method,
    });
    if (req.url.endsWith("/project")) {
      res.writeHead(307, { Location: "/no" });
      res.end();
      return;
    }
    res.setHeader("Content-Type", "application/json");
    res.end(
      JSON.stringify({ id, email: "test@agent.example.com", name: "Test" }),
    );
  });
  await new Promise((resolve) => server.listen(0, "127.0.0.1", resolve));
  t.after(() => {
    server.closeAllConnections();
    server.close();
  });
  const baseUrl = `http://127.0.0.1:${server.address().port}`,
    key = "exp_" + "p".repeat(43);
  const project = new ExpectboxProject({ apiKey: key, baseUrl });
  assert.throws(() => new ExpectboxAgent({ apiKey: key }), /key/i);
  assert.throws(
    () => new ExpectboxProject({ apiKey: "exa_" + "a".repeat(43) }),
    /key/i,
  );
  assert.ok(!JSON.stringify(project).includes(key));
  for (let n = 0; n < 2; n++)
    await project.createInbox({ name: "Test", username: "test" }, "saved-key");
  assert.equal(received[0].url, "/api/agents/v1/project/inboxes");
  assert.equal(received[0].headers["idempotency-key"], "saved-key");
  assert.equal(received[0].body, received[1].body);
  await project.createInboxKey(id, {
    name: "Runtime",
    scopes: ["messages:read"],
  });
  await project.revokeInboxKey(id, id);
  assert.equal(received[3].method, "DELETE");
  assert.throws(() => project.createInbox({}, ""), /idempotency/);
  assert.throws(() => project.createInboxKey("../../x", {}), /UUID/);
  await assert.rejects(project.project());
  assert.equal(received.length, 5);
  assert.ok(!received.some((r) => r.url === "/no"));
  for (const origin of [
    "http://example.com",
    "https://a:b@example.com",
    "https://example.com/api",
    "https://example.com/#x",
  ])
    assert.throws(() => new ExpectboxProject({ apiKey: key, baseUrl: origin }));
});
